### PR TITLE
Move swiftpm to swift_build_support infra

### PR DESCRIPTION
**Pull request by aciidgh**
We should have done this ages ago. This will allow cleaning up most of
the hacks in SwiftPM's build script.

Dependencies:

- https://github.com/apple/swift/pull/27580

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
This means we can rip swiftpm out of build-script-impl but I need to wait until @ahoppen migrates SwiftPM build-script-impl dependents.

**Comment by aciidb0mb3r** [comment]
@swift-ci smoke test

File: utils/swift_build_support/swift_build_support/products/swiftpm.py [hunk 1]
```diff
@@ -10,8 +10,65 @@
...
+        product.source_dir, 'Utilities', 'bootstrap')
+    toolchain_path = args.install_destdir
+    if platform.system() == 'Darwin':
+        # The prefix is an absolute path, so concatenate without os.path.
+        toolchain_path += \
+            product.targets.darwin_toolchain_prefix(args.install_prefix)
+
+    release = '' if args.build_variant == 'Debug' else '--release'
+
+    swiftc = os.path.join(toolchain_path, "usr", "bin", "swiftc")
```
Comment (ahoppen) [inline utils/swift_build_support/swift_build_support/products/swiftpm.py]: IIUC the idea is that you only pass the toolchain directory to SwiftPM's build script and it will find `swiftc` in there on its own. See https://github.com/apple/swift/pull/27580#pullrequestreview-300222128

**Comment by aciidb0mb3r** [comment]
Dependencies:

- https://github.com/apple/swift/pull/27580

File: utils/swift_build_support/swift_build_support/products/swiftpm.py [hunk 2]
```diff
@@ -2,16 +2,72 @@
...
+
+    @classmethod
+    def is_build_script_impl_product(cls):
+        return False
+
+    def build(self, host_target):
+        run_build_script_helper('build', host_target, self, self.args)
+
+    def test(self, host_target):
+        run_build_script_helper('test', host_target, self, self.args)
```
Comment (ahoppen) [inline utils/swift_build_support/swift_build_support/products/swiftpm.py]: You should not run the tests unless your test flag is passed. See 
https://github.com/apple/swift/pull/27580/files#diff-bac1aa9ca5d00ef87144461e6bc764edR77

(this should actually be changed that the driver takes care of checking the flag)

File: utils/swift_build_support/swift_build_support/products/swiftpm.py [hunk 3]
```diff
@@ -2,16 +2,72 @@
...
+
+    def install(self, host_target):
+        run_build_script_helper('install', host_target, self, self.args)
```
Comment (ahoppen) [inline utils/swift_build_support/swift_build_support/products/swiftpm.py]: Same as above. Check that your install flag is set.

File: utils/swift_build_support/swift_build_support/products/swiftpm.py [hunk 4]
```diff
@@ -2,16 +2,72 @@
...
+
+def run_build_script_helper(action, host_target, product, args):
+    script_path = os.path.join(
+        product.source_dir, 'Utilities', 'bootstrap')
+    toolchain_path = args.install_destdir
+    if platform.system() == 'Darwin':
+        # The prefix is an absolute path, so concatenate without os.path.
+        toolchain_path += \
+            product.targets.darwin_toolchain_prefix(args.install_prefix)
```
Comment (ahoppen) [inline utils/swift_build_support/swift_build_support/products/swiftpm.py]: Not part of this PR but: We should factor this building of the toolchain path out to the driver and not copy it into every PR. I’ll take care of it

File: utils/swift_build_support/swift_build_support/products/swiftpm.py [hunk 5]
```diff
@@ -2,16 +2,72 @@
...
+
+    swiftc = os.path.join(toolchain_path, "usr", "bin", "swiftc")
+    sbt = os.path.join(toolchain_path, "usr", "bin", "swift-build-tool")
```
Comment (ahoppen) [inline utils/swift_build_support/swift_build_support/products/swiftpm.py]: Same as above. SwiftPM’s build script should find the `swift-build-tool`

**Review by ahoppen — COMMENTED** [review]
Thanks for tackling this.

Some comments based on an initial review inline.

File: utils/swift_build_support/swift_build_support/products/swiftpm.py (hunk 2 above)
Comment (aciidb0mb3r) [reply utils/swift_build_support/swift_build_support/products/swiftpm.py]: huh, ok.

File: utils/swift_build_support/swift_build_support/products/swiftpm.py (hunk 4 above)
Comment (aciidb0mb3r) [reply utils/swift_build_support/swift_build_support/products/swiftpm.py]: yeah

File: utils/swift_build_support/swift_build_support/products/swiftpm.py (hunk 1 above)
Comment (aciidb0mb3r) [reply utils/swift_build_support/swift_build_support/products/swiftpm.py]: I agree but I'll do that later since SwiftPM's bootstrap script is a hot mess (which I am [reworking](https://github.com/apple/swift-package-manager/blob/master/Utilities/new-bootstrap) with the new Swift support in CMake). It'll be much easier to do this in the new script vs try to fight the old one.

**Comment by aciidb0mb3r** [comment]
@swift-ci smoke test

**Comment by aciidb0mb3r** [comment]
@swift-ci smoke test

**Comment by DougGregor** [comment]
Is this how we would integrate [swift-driver](https://github.com/apple/swift-driver) into `build-script`?

**Comment by aciidb0mb3r** [comment]
@DougGregor Yeah, it would be a good start to start building it after swiftpm. We'll need to think of an alternative ways once we actually want to build it before swiftpm.

File: utils/swift_build_support/swift_build_support/products/swiftpm.py (hunk 1 above)
Comment (ahoppen) [reply utils/swift_build_support/swift_build_support/products/swiftpm.py]: Sounds good.
